packages/product-json-to-csv/test/main.spec.js

```diff
@@ -0,0 +1,625 @@
+        projectKey: 'foo',
+      }
+      const defaultConfig = {
+        delimiter: ',',
+        multiValueDelimiter: ';',
+        fillAllRows: false,
+        categoryBy: 'name',
+        categoryOrderHintBy: 'name',
+        language: 'en',
+        productSeparator: '\n',
+      }
+
+      productJsonToCsv = new ProductJsonToCsv(apiConfig)
+      expect(productJsonToCsv.logger).toBeDefined()
+      expect(productJsonToCsv.client).toBeDefined()
+      expect(productJsonToCsv.parserConfig).toEqual(defaultConfig)
+    })
+  })
+
+  describe('::run', () => {
+    let zipSpy
+    let csvSpy
+    beforeEach(() => {
+      zipSpy = jest.spyOn(writer, 'writeToZipFile').mockImplementation(() => {})
+      csvSpy = jest
+        .spyOn(writer, 'writeToSingleCsvFile')
+        .mockImplementation(() => {})
+    })
+
+    it('should write data to single `csv` file if headers are set', () => {
+      productJsonToCsv.parse = jest.fn(() => 'foo')
+      productJsonToCsv.parserConfig.headers = []
+
+      productJsonToCsv.run()
+      expect(csvSpy).toBeCalled()
+      expect(zipSpy).not.toBeCalled()
+    })
+
+    it('should write data to `zip` file if headers are not set', () => {
+      productJsonToCsv.parse = jest.fn(() => 'bar')
+
+      productJsonToCsv.run()
+      expect(csvSpy).not.toBeCalled()
+      expect(zipSpy).toBeCalled()
+    })
+  })
+
+  describe('::parse', () => {
+    let product
+    let inputStream
+    beforeEach(() => {
+      productJsonToCsv._resolveReferences = jest.fn(data =>
+        Promise.resolve(data)
+      )
+      productJsonToCsv.logger.error = jest.fn()
+      product = oneLineTrim`
+        {
+          "id": "product-1-id",
+          "slug": {
+            "en": "my-slug-1"
+          },
+          "masterVariant": {
+            "id": "mv-id",
+            "key": "mv-key"
+          },
+          "variants": []
+        }`
+      inputStream = streamTest.fromChunks([product])
+    })
+
+    it('should take an inputStream and output highland stream', () => {
+      const productStream = productJsonToCsv.parse(inputStream)
+      expect(productStream.source.__HighlandStream__).toBeTruthy()
+    })
+
+    it('should return a flattened product object', done => {
+      const productStream = productJsonToCsv.parse(inputStream)
+      const expected = [
+        {
+          id: 'product-1-id',
+          'slug.en': 'my-slug-1',
+          'variant.id': 'mv-id',
+          'variant.key': 'mv-key',
+        },
+      ]
+      productStream.toArray(actual => {
+        expect(actual).toEqual(expect.arrayContaining(expected))
+        done()
+      })
+    })
+
+    it('should log and emit error if error occurs', done => {
+      const fakeError = new Error('fake error')
+      productJsonToCsv._resolveReferences = jest.fn(() =>
+        Promise.reject(fakeError)
+      )
+      const outputStream = {
+        emit: jest.fn(),
+      }
+      const productStream = productJsonToCsv.parse(inputStream, outputStream)
```

Comment:
Maybe just flavour but we put setup into its. Where BDD advocates that to be more in describes and a beforeEach  

describe: when erros
   beforeEach
   it should log